### ISSUE TITLE
CPU ai team discuss here

### ISSUE THREAD

**Issue by motoko-11** [OpenC-IIIT/prosfair#10]

**Comment by arshadsid** [comment]
Our code currently treats check mate and check as equal 
So for our cpu 
Having a check mate is same as having check.

**Comment by motoko-11** [comment]
Yeah. Plus the initial state of the king when it's surrounded by it's own player also results in checkmate.. though I will soon commit a solution to it... it's trivial.. WE need better evaulation function..

**Comment by arshadsid** [comment]
How will same team pieces result in checkmate we are comparing opposite team color in the moves right?

**Comment by motoko-11** [comment]
Think about it... It's basically how I wrote the for loop which creates such situation... what is there is no move possible for the king.. it will directly return checkmate

**Comment by arshadsid** [comment]
Ok yaa..
So if its not check them the king dosent have to move either
In this case the king has to move even if not check

**Comment by sk364** [comment]
We kind of debugged that. But, we didn't commit.

**Comment by lordzuko** [comment]
@motoko-11  How is the evaluate board to generate the scores for intermediate chess pieces ??

**Comment by motoko-11** [comment]
evaluate board is very priliminary.. it is only considering check and checkmate situations.. in real it should return values which are proportional to the number of pieces one player has against another... kind of stuff....@lordzuka

**Comment by lordzuko** [comment]
ya i know that , all I am asking is how are we going to give the score for a current position ??

**Comment by motoko-11** [comment]
the most basic one is  white_pieces * priority - black_pieces * priority , prioirty is written in offence.py ... and this is what is followed mostly...at a initial stage..

**Comment by lordzuko** [comment]
but what i am thinking is we should more concentrate on the position in the move will leave the system (i mean game) rather than assigning priority so ....

I would suggest some thing like 

**get_scorefun(board,color){
 **   sum = 0;
 **  for pieces on board[color]{
 **       sum += board[color][piece]*piece_priority; 
  **  }
  **  return sum;
**}
**score = getscorefun(board,"white") - getscorefun(board,"black")


what do u think about this ??

**Comment by motoko-11** [comment]
Where you have not assinged priority... I meant exactly the same thing... @lordzuko ... Evaluation function evaluates the goodness of the board from point of view of one player... that's exactly how I was imagining it should be added to it. :) ...

**Comment by lordzuko** [comment]
cool so few things has been decided .. :+1:

**Comment by lordzuko** [comment]
game doesn't stops even after king is dead ...!!!

**Comment by motoko-11** [comment]
Actually I didn't make it to. I am not actually trying to make traditional chess. As soon as design gets more clear, I will start to use this code as a template. Although If you can check, kindly review the use of map, since there is a bug which occurs while playing sometimes due to keyError.

**Comment by lordzuko** [comment]
which map ??

**Comment by motoko-11** [comment]
the chessboard map...which represents the chessboard pieces.

**Comment by lordzuko** [comment]
u mean the json file of chess pieces which i reformatted ?? ...

**Comment by motoko-11** [comment]
Noo... I am saying the way we have used python maps... I think there is a logical bug somewhere in the code.

**Comment by lordzuko** [comment]
ya that's suppose to be, as I was telling u using maps as only structure to board has lots of disadvantages:
1) difficult to find indexes directly
2) difficult to understand 
3) difficult to add more information per square without changing the data structure

so what we can do is create simple structure of board square and start working on it.

**Comment by sk364** [comment]
How to use alpha beta pruning with minimax or alpha beta pruning is used alone?

**Comment by sk364** [comment]
The code for alpha beta pruning - 

def alphaBetaMin(board,color,alpha,beta, depth):
        if depth == 0 or helper.game_over(board, color) :
                return evaluate_board(board,color)

        moves_list = helper.get_moves(board,color)

        for move in moves_list:
                clone_board = helper.generate_board(board,move)

                score = alphaBetaMax(clone_board,opposite_army[color],alpha,beta,depth-1)
                if score <= beta:
                        return alpha
                if score < alpha:
                        beta = score

        return beta

def alphaBetaMax(board,color,alpha,beta, depth):
        if depth == 0 or helper.game_over(board, color) :
                return evaluate_board(board,color)

        moves_list = helper.get_moves(board,color)

        for move in moves_list:
                clone_board = helper.generate_board(board,move)

                score = alphaBetaMin(clone_board,opposite_army[color],alpha,beta,depth-1)
                if score >= beta:
                        return beta
                if score > alpha:
                        alpha = score

        return alpha

**Comment by sk364** [comment]
why is the cpp alpha beta pruning code not working with cpu vs cpu mode? I don't seem to get what is the problem...!
